ikarus/finiteElements/mechanics/nonLinearElasticityFE.hh

```diff
@@ -28,112 +30,240 @@
+        : BasePowerFE(globalBasis.flat(), element),
+          volumeLoad{Std::returnReferenceOrNulloptIfObjectIsNullPtr(p_volumeLoad)},
+          neumannBoundaryLoad_{Std::returnReferenceOrNulloptIfObjectIsNullPtr(p_neumannBoundaryLoad)},
           neumannBoundary_{neumannBoundary},
           mat{p_mat} {
       this->localView().bind(element);
+      auto& first_child = this->localView().tree().child(0);
+      const auto& fe    = first_child.finiteElement();
+      numberOfNodes     = fe.size();
+      dispAtNodes.resize(fe.size());
       const int order = 2 * (this->localView().tree().child(0).finiteElement().localBasis().order());
       localBasis      = Dune::CachedLocalBasis(this->localView().tree().child(0).finiteElement().localBasis());
       localBasis.bind(Dune::QuadratureRules<double, Traits::mydim>::rule(this->localView().element().type(), order),
                       Dune::bindDerivatives(0, 1));
+      assert(((not neumannBoundary_ and not neumannBoundaryLoad_) or (neumannBoundary_ and neumannBoundaryLoad_))
+             && "If you pass a Neumann boundary you should also pass the function for the Neumann load!");
     }
 
-    using Traits = TraitsFromLocalView<LocalView>;
+  public:
+    auto getDisplacementFunction(const FERequirementType& par) const {
+      const auto& d = par.getGlobalSolution(Ikarus::FESolutions::displacement);
 
-  private:
-    template <class ScalarType>
-    ScalarType calculateScalarImpl(const FERequirementType& req, Eigen::VectorX<ScalarType>& dx) const {
-      const auto& d      = req.getGlobalSolution(Ikarus::FESolutions::displacement);
-      const auto& lambda = req.getParameter(Ikarus::FEParameter::loadfactor);
-      using namespace Dune::DerivativeDirections;
-      using namespace Dune;
-      auto& first_child = this->localView().tree().child(0);
-      const auto& fe    = first_child.finiteElement();
-      Dune::BlockVector<RealTuple<ScalarType, Traits::dimension>> disp(fe.size());
+      for (auto i = 0U; i < dispAtNodes.size(); ++i)
+        for (auto k2 = 0U; k2 < myDim; ++k2)
+          dispAtNodes[i][k2] = d[this->localView().index(this->localView().tree().child(k2).localIndex(i))[0]];
 
-      for (auto i = 0U; i < fe.size(); ++i)
-        for (auto k2 = 0U; k2 < Traits::mydim; ++k2)
-          disp[i][k2]
-              = dx[i * 2 + k2] + d[this->localView().index(this->localView().tree().child(k2).localIndex(i))[0]];
+      auto geo = std::make_shared<const typename GridView::GridView::template Codim<0>::Entity::Geometry>(
+          this->localView().element().geometry());
+      Dune::StandardLocalFunction uFunction(localBasis, dispAtNodes, geo);
+      return uFunction;
+    }
+    auto getStrainFunction(const FERequirementType& par) const {
+      return greenLagrangeStrains(getDisplacementFunction(par));
+    }
 
-      ScalarType energy = 0.0;
+    auto getMaterialTangent(const auto& strain, const bool& voigt = true) const {

```

Comment:
The boolean needs to be a template parameter, since in the call `tangentModuli<strainType, voigt>` it is also a template parameter.